### PR TITLE
Pytest unit tests, new losses support, and normalization enhancement

### PR DESCRIPTION
**Pull request by franckma31**
The Torchlip library has undergone several updates to align more closely with the functionalities provided by deel-lip (TensorFlow). Here are the key changes:

- Transition to pytest for unit tests, sharing a common test suite with deel-lip. The utils_framework.py file has been introduced to facilitate adaptation of names and parameters.

- Enhancements in the available loss functions:

  - Binary and Multiclass KR and HKR:  now support any target style (0,1), (-1,1), ...  [target>0 for true value], 
  - Multi-GPU Support: binary and multiclass KR and HKR supports multi-gpu mode with preprocessed target
  - Modified HKR Loss: the $\alpha$ parameter (required in  [0,1]) with (1- $\alpha$ )HKRLoss+ $\alpha$ HingeLoss (for backward compatibility a warning is raised in case of $\alpha>1$ replacing the given value by $\frac{\alpha}{1+\alpha}$)
  - Tau Cross-Entropy Losses: Support has been added for cross-entropy (binary or categorical) losses with temperature scaling.
  - Reduction Parameters: Loss functions now accept the 'reduction' argument with options: 'mean', 'none', or 'sum'.
-  Normalizers (Spectral and Bjorck): Updated with a stopping criterion based on an epsilon value (eps) rather than a fixed number of iterations
- Bjorck parenthesis trick added (reduce computational complexity)

### CODE REVIEW

File: tests/utils_framework.py [hunk 1]
```diff
@@ -553,9 +549,9 @@ def scaleDivAlpha(alpha):
 def SpectralInitializer(eps_spectral, eps_bjorck):
     warnings.warn("spectral_ and bjorck_ require n_iterations in pytorch")
```
Comment (cofri) [inline tests/utils_framework.py:550]: Is this warning still required with the new `eps_spectral` and `eps_bjorck`?

File: tests/test_condense.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        steps_per_epoch=steps_per_epoch,
+        callbacks=callback_list,
+    )
+    # model.__getattribute__(FIT)(
+    #     linear_generator(batch_size, input_shape, kernel),
+    #     steps_per_epoch=steps_per_epoch,
+    #     epochs=epochs,
+    #     verbose=0,
+    #     callbacks=callback_list,
+    # )
```
Comment (cofri) [inline tests/test_condense.py:245]: We can remove the code from TensorFlow/Keras in comments at different places in this file

**Review by cofri — COMMENTED** [review]
So many valuable changes! Nice and big work

File: deel/torchlip/modules/module.py [hunk 3]
```diff
@@ -29,16 +29,68 @@
...
+    is 1-Lipschitz for specific set of parameters.
+    """
+    supported_1lip_layers = (
+        torch.nn.Softmax,
+        torch.nn.Flatten,
+        torch.nn.Identity,
+        torch.nn.ReLU,
+        torch.nn.Sigmoid,
+        torch.nn.Tanh,
+        Reshape,
```
Comment (massena-t) [inline deel/torchlip/modules/module.py:57]: add torch.view to the list ?

File: deel/torchlip/modules/loss.py [hunk 4]
```diff
@@ -161,3 +160,139 @@ def __init__(
...
+            alpha (float): regularization factor (0 <= alpha <= 1),
+                0 for KR only, 1 for hinge only
+            min_margin (float): margin to enforce.
+            temperature (float): factor for softmax  temperature
+                (higher value increases the weight of the highest non y_true logits)
+            alpha_mean (float): geometric mean factor
+            one_hot_ytrue (bool): set to True when y_true are one hot encoded (0 or 1),
+                and False when y_true already signed bases (for instance +/-1)
+            reduction: passed to tf.keras.Loss constructor
+            name (str): passed to tf.keras.Loss constructor
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/torchlip/modules/loss.py]: the commit 7592b906d7c6e49d6f1e00205ed382da02f0989a solved the issue

File: deel/torchlip/modules/loss.py [hunk 5]
```diff
@@ -35,18 +35,20 @@ class KRLoss(torch.nn.Module):
     duality.
     """
 
-    def __init__(self, true_values=None):
+    def __init__(self, reduction: str = "mean", true_values=None):
         """
         Args:
             true_values: tuple containing the two label for each predicted class.
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/torchlip/modules/loss.py]: the commit https://github.com/deel-ai/deel-torchlip/commit/7592b906d7c6e49d6f1e00205ed382da02f0989a solved the issue

File: deel/torchlip/modules/loss.py [hunk 6]
```diff
@@ -343,3 +346,86 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
             target = torch.Tensor(target, dtype=input.dtype)
         loss_batch = self.fct(target, input)
         return F.apply_reduction(loss_batch, self.reduction)
+
+
+class TauCrossEntropyLoss(CrossEntropyLoss):
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/torchlip/modules/loss.py:428]: the commit https://github.com/deel-ai/deel-torchlip/commit/7592b906d7c6e49d6f1e00205ed382da02f0989a solved the issue

File: tests/test_condense.py (hunk 2 above)
Comment (franckma31) [reply tests/test_condense.py]: the commit 26fd0e30791e956a135fce6f1ea71d3c02ccbd05 solved this issue

File: deel/torchlip/modules/module.py (hunk 3 above)
Comment (franckma31) [reply deel/torchlip/modules/module.py:57]: torch.tensor.view is not a layer

File: tests/utils_framework.py (hunk 1 above)
Comment (franckma31) [reply tests/utils_framework.py]: the commit e83741e452cac91e7f6a7713f060443199b1c97f solved this issue

File: deel/torchlip/modules/module.py [hunk 7]
```diff
@@ -163,3 +183,12 @@ def vanilla_export(self):
...
         return TorchSequential(OrderedDict(layers))
+
+
+class Reshape(torch.nn.Module):
+    def __init__(self, target_shape):
+        super(Reshape, self).__init__()
+        self.target_shape = target_shape
+
+    def forward(self, x):
+        return reshape(x, self.target_shape)
```
[the remark this answers is not in the document]
Comment (franckma31) [reply deel/torchlip/modules/module.py]: Since the Reshape module was only used in pytest, it has been replaced by a torch.nn.UnFlatten layer. This code has been removed.

File: deel/torchlip/modules/module.py (hunk 3 above)
Comment (franckma31) [reply deel/torchlip/modules/module.py]: But the module  torch.nn.UnFlatten has been added to the list in the commit 9c2d4a126ef33f992e246d4eeacbb0cd803c1a5f

**Review by cofri — APPROVED** [review]

**Review by cofri — APPROVED** [review]
